#### libraries used
1. Pandas is a powerful library for data manipulation and analysis
2. NumPy is a fundamental package for numerical computations in Python
3. A high-level interface for creating interactive visualizations. Plotly Express simplifies the process of creating a variety of plots 

In [1]:
#import the  libraries
import pandas as pd
import numpy as np
import plotly as px
import warnings            
warnings.filterwarnings("ignore")


In [2]:
#import the dataset 
df = pd.read_csv("globalterrorismdb_0718dist.csv", encoding="ISO-8859-1")
df.head()

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaN       58   
1  197000000002   1970       0     0        NaN         0        NaN      130   
2  197001000001   1970       1     0        NaN         0        NaN      160   
3  197001000002   1970       1     0        NaN         0        NaN       78   
4  197001000003   1970       1     0        NaN         0        NaN      101   

          country_txt  region  ... addnotes scite1 scite2  scite3  dbsource  \
0  Dominican Republic       2  ...      NaN    NaN    NaN     NaN      PGIS   
1              Mexico       1  ...      NaN    NaN    NaN     NaN      PGIS   
2         Philippines       5  ...      NaN    NaN    NaN     NaN      PGIS   
3              Greece       8  ...      NaN    NaN    NaN     NaN      PGIS   
4               Japan       4  ...      NaN    NaN    NaN     NaN      PGIS   

   INT_LOG  INT_IDEO INT_MISC INT_ANY  related  
0        0         0        0       0      NaN  
1        0         1        1       1      NaN  
2       -9        -9        1       1      NaN  
3       -9        -9        1       1      NaN  
4       -9        -9        1       1      NaN  

[5 rows x 135 columns]

In [3]:
df=df[["iyear","imonth","iday","country_txt","region_txt","provstate","city",
       "latitude","longitude","location","summary","attacktype1_txt","targtype1_txt",
       "gname","motive","weaptype1_txt","nkill","nwound","addnotes"]]
df.head()

iyear  imonth  iday         country_txt                   region_txt  \
0   1970       7     2  Dominican Republic  Central America & Caribbean   
1   1970       0     0              Mexico                North America   
2   1970       1     0         Philippines               Southeast Asia   
3   1970       1     0              Greece               Western Europe   
4   1970       1     0               Japan                    East Asia   

  provstate           city   latitude   longitude location summary  \
0       NaN  Santo Domingo  18.456792  -69.951164      NaN     NaN   
1   Federal    Mexico city  19.371887  -99.086624      NaN     NaN   
2    Tarlac        Unknown  15.478598  120.599741      NaN     NaN   
3    Attica         Athens  37.997490   23.762728      NaN     NaN   
4   Fukouka        Fukouka  33.580412  130.396361      NaN     NaN   

                  attacktype1_txt                targtype1_txt  \
0                   Assassination  Private Citizens & Property   
1     Hostage Taking (Kidnapping)      Government (Diplomatic)   
2                   Assassination          Journalists & Media   
3               Bombing/Explosion      Government (Diplomatic)   
4  Facility/Infrastructure Attack      Government (Diplomatic)   

                                gname motive weaptype1_txt  nkill  nwound  \
0                              MANO-D    NaN       Unknown    1.0     0.0   
1  23rd of September Communist League    NaN       Unknown    0.0     0.0   
2                             Unknown    NaN       Unknown    1.0     0.0   
3                             Unknown    NaN    Explosives    NaN     NaN   
4                             Unknown    NaN    Incendiary    NaN     NaN   

  addnotes  
0      NaN  
1      NaN  
2      NaN  
3      NaN  
4      NaN

In [4]:
df.rename(columns={"iyear":"Year","imonth":"Month","iday":"Day","country_txt":"Country",
                   "region_txt":"Region","provstate":"Province/State","city":"City",
                   "latitude":"Latitude","longitude":"Longitude","location":"Location",
                   "summary":"Summary","attacktype1_txt":"Attack Type","targtype1_txt":"Target Type",
                   "gname":"Group Name","motive":"Motive","weaptype1_txt":"Weapon Type",
                   "nkill":"Killed","nwound":"Wounded","addnotes":"Add Notes"},inplace=True)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181691 entries, 0 to 181690
Data columns (total 19 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Year            181691 non-null  int64  
 1   Month           181691 non-null  int64  
 2   Day             181691 non-null  int64  
 3   Country         181691 non-null  object 
 4   Region          181691 non-null  object 
 5   Province/State  181270 non-null  object 
 6   City            181257 non-null  object 
 7   Latitude        177135 non-null  float64
 8   Longitude       177134 non-null  float64
 9   Location        55495 non-null   object 
 10  Summary         115562 non-null  object 
 11  Attack Type     181691 non-null  object 
 12  Target Type     181691 non-null  object 
 13  Group Name      181691 non-null  object 
 14  Motive          50561 non-null   object 
 15  Weapon Type     181691 non-null  object 
 16  Killed          171378 non-null  float64
 17  Wounded   

In [6]:
df.isnull().sum()

Year                   0
Month                  0
Day                    0
Country                0
Region                 0
Province/State       421
City                 434
Latitude            4556
Longitude           4557
Location          126196
Summary            66129
Attack Type            0
Target Type            0
Group Name             0
Motive            131130
Weapon Type            0
Killed             10313
Wounded            16311
Add Notes         153402
dtype: int64

In [7]:
df.describe()

Year          Month            Day       Latitude  \
count  181691.000000  181691.000000  181691.000000  177135.000000   
mean     2002.638997       6.467277      15.505644      23.498343   
std        13.259430       3.388303       8.814045      18.569242   
min      1970.000000       0.000000       0.000000     -53.154613   
25%      1991.000000       4.000000       8.000000      11.510046   
50%      2009.000000       6.000000      15.000000      31.467463   
75%      2014.000000       9.000000      23.000000      34.685087   
max      2017.000000      12.000000      31.000000      74.633553   

          Longitude         Killed        Wounded  
count  1.771340e+05  171378.000000  165380.000000  
mean  -4.586957e+02       2.403272       3.167668  
std    2.047790e+05      11.545741      35.949392  
min   -8.618590e+07       0.000000       0.000000  
25%    4.545640e+00       0.000000       0.000000  
50%    4.324651e+01       0.000000       0.000000  
75%    6.871033e+01       2.000000       2.000000  
max    1.793667e+02    1570.000000    8191.000000

In [8]:
import plotly.graph_objs as go
# Calculate the number of attacks per year
attacks = df["Year"].value_counts(dropna=False).sort_index().to_frame().reset_index().rename(columns={"index": "Year", "Year": "Attacks"})

color_scale = px.colors.sequential.Viridis
colors = [color_scale[int(val / max(attacks['Attacks']) * (len(color_scale) - 1))] for val in attacks['Attacks']]

# Create the bar chart 
fig = go.Figure(data=[ go.Bar(x=attacks['Year'], y=attacks['Attacks'], marker=dict(color=colors))])
fig.update_layout(title='Timeline of Attacks', xaxis_title='Years', yaxis_title='Number of Attacks')

fig.show()

### From the chart we could say that, in 2014 more number of Attacks have taken place

In [9]:
# Calculate the number of people killed and wounded per year
Year_wounded = df[["Year", "Wounded"]].groupby("Year").sum()
Year_killed = df[["Year", "Killed"]].groupby("Year").sum()

# Create the subplot 
fig = go.Figure()

# "People Killed" bar chart subplot
fig.add_trace(go.Bar(x=Year_killed.index, y=Year_killed['Killed'], name='People Killed'))

# Add "People Wounded" bar chart subplot
fig.add_trace(go.Bar(x=Year_wounded.index, y=Year_wounded['Wounded'], name='People Wounded'))

fig.update_layout(title='People Killed and Wounded in each Year',
                  xaxis_title='Years', yaxis_title='Number of People',
                  barmode='stack')  # Stacked bar chart


fig.show()

In [10]:
region = pd.crosstab(df.Year, df.Region)


fig = go.Figure()

for col in region.columns:
    fig.add_trace(go.Scatter(x=region.index, y=region[col], mode='lines', name=col, fill='tonexty'))

# Update layout and labels
fig.update_layout(title='Region wise attacks',
                  xaxis_title='Years', yaxis_title='Number of Attacks',
                  showlegend=True, template='plotly')

fig.show()

In [11]:
# Create a cross-tabulation of attacks by year and region
reg = pd.crosstab(df.Year, df.Region)

# Transpose the cross-tabulation to get regions as columns
regt = reg.transpose()

# Calculate the total attacks for each region
regt['Total'] = regt.sum(axis=1)

# Sort regions by the total number of attacks
ra = regt['Total'].sort_values(ascending=False)

# Create the bar chart using Plotly
fig = go.Figure(data=[
    go.Bar(x=ra.index, y=ra, marker_color=['red', 'orange', 'yellow', 'green', 'blue', 'purple'])
])
fig.update_layout(title='Total Number of Attacks in each Region from 1970-2017',
                  xaxis_title='Region', yaxis_title='Number of Attacks')


fig.show()

In [13]:
# Calculate the total number of people killed and wounded for each region
Region_Killed = df[["Region", "Killed"]].groupby("Region").sum().sort_values(by="Killed", ascending=False)
Region_Wounded = df[["Region", "Wounded"]].groupby("Region").sum().sort_values(by="Wounded", ascending=False)

# Create the bar chart for People Killed using Plotly Express
fig_killed = px.bar(Region_Killed, x=Region_Killed.index, y='Killed', title='People Killed in each Region', color='Killed')
fig_killed.update_layout(xaxis_title='Regions', yaxis_title='Number of People Killed')

# Create the bar chart for People Wounded using Plotly Express
fig_wounded = px.bar(Region_Wounded, x=Region_Wounded.index, y='Wounded', title='People Wounded in each Region', color='Wounded')
fig_wounded.update_layout(xaxis_title='Regions', yaxis_title='Number of People Wounded')


fig_killed.show()
fig_wounded.show()

In [14]:
# Calculate the count of terrorist attacks per country
country_attacks = df.groupby('Country').size().reset_index(name='count').sort_values(by='count', ascending=False)
top5_country = country_attacks.head(5)

# Create the bar chart using Plotly Express
fig = px.bar(top5_country, x='Country', y='count', title='Top 5 countries with the most terrorist attacks from 1970 to 2017',
             color='count', color_continuous_scale='viridis')
fig.update_layout(xaxis_title='Countries', yaxis_title='Number of Terrorist Attacks')

fig.show()

In [15]:
# Calculate the distribution of attack types
attack_type_counts = df['Attack Type'].value_counts()

fig = px.pie(attack_type_counts, values=attack_type_counts.values, names=attack_type_counts.index,
             title='Attack Type Distribution in Terrorism Dataset', labels={'label': 'Attack Type'})

fig.show()

In [16]:
# Calculate the distribution of target types
target_type_counts = df['Target Type'].value_counts()

fig = px.pie(target_type_counts, values=target_type_counts.values, names=target_type_counts.index,
             title='Target Type Distribution in Terrorism Dataset', labels={'label': 'Target Type'},width=1100, height=800)

fig.show()

In [17]:
attack_type_counts = df['Attack Type'].value_counts()

# Calculate percentage values
total_attacks = len(df)
attack_type_percentages = (attack_type_counts / total_attacks) * 100

# Create a dataframe for the treemap
treemap_data = pd.DataFrame({'Attack Type': attack_type_counts.index, 'Percentage': attack_type_percentages.values})

# Create the treemap using Plotly Express
fig = px.treemap(treemap_data, path=['Attack Type'], values='Percentage', 
                 title='Attack Type Distribution in Terrorism Dataset',width=1100, height=800)

fig.show()

### From the Explored data,
1. highest number of Terrorist Attacks happendn in contry : Iraq
2. Regions with the highest number of Terrorist Attacks happend in Region: Middle East & North Africa
3. Maximum number of people were killed:1570 people
4. Year with the most Attacks: 2014
5. Group with the most Attacks: Taliban
6. Most used Attacks Types: Bombing/Explosion